Columns block: Enable blockGap and vertical margin support

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Following on from #33991 and the Columns CSS change by @richtabor in https://github.com/WordPress/gutenberg/pull/34456, this PR opts the Columns block in to the gap block support control, and also the vertical margin control.

Adjusting block gap allows the user to adjust the space between columns blocks. Depending on where the columns block is placed, and whether or not the theme uses the Layout support, the adjusted block gap value can also change the gap between the selected block and blocks around it. To allow the user to tweak this, the margin control is added, so that the margin can be controlled separately to the gap on inner blocks.

***Up for discussion:*** which controls are good to display by default? Does it make sense to expose blockGap by default, but hide margin? Should we default to both? I'd love to hear feedback on what's best to expose here.

### Changes proposed

* Opt the Columns block in to using the blockGap support
* Opt the Columns block in to using the margin support for vertical margins
* Update the gap support label to `Block spacing`

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Running TT1-Blocks theme, in the `theme.json` file, under `settings.spacing`, enable the `blockGap` and `customMargin` controls. For example, here's how that section of my `theme.json` file looks:

```
		"spacing": {
			"blockGap": true,
			"customPadding": true,
			"customMargin": true,
			"units": [
				"px",
				"em",
				"rem",
				"vh",
				"vw"
			]
		},
```

1. Add a Columns block to a post or page, and try out the blockGap control in the inspector controls.
2. Try adding the Columns block to different positions in a page or post, and adjust the blockGap value.
3. In the Dimensions panel, enable the Margin control, and try adjusting the margin to adjust the spacing around the block.
4. Test in the site editor

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/14988353/132447637-a05f5e56-6f35-4d49-b6f4-044cc1ee9ad4.mp4

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested. (manually)
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Columns block: Enable blockGap and vertical margin support

andrewserong commented:
Joen, I've added you as a reviewer since we've both been working on gap support, please let me know if there are other folks I should ping for feedback, too!

jasmussen commented:
Always happy to review 👌 

Nice work on this one. I like that it's a minimal change to the codebase. More than anything this illustrates the need for some thought put into the sidebar panels. We've been talking about Dimensions, Spacing and Layout panels for a bit now, and waffling on which controls go where, and whether perhaps we have only two of the above. For example, gap might be a _layout_ control since it affects the items inside. But then isn't spacing? And if we do have spacing, couldn't block gap be part of that? I know this is something @youknowriad has thought a lot about as well, would love his input.

richtabor commented:
> For example, gap might be a _layout_ control since it affects the items inside. But then isn't spacing? And if we do have spacing, couldn't block gap be part of that? I know this is something @youknowriad has thought a lot about as well, would love his input.

Totally agree, and something I started highlighting in https://github.com/WordPress/gutenberg/issues/34558. I feel we could condense most of the dimensions type controls into a more global Layout panel, but the tricky part is, that while the Group block has a "Layout" panel, the "Columns" block does not. 

Comes down to the idea that perhaps we have a Layout panel that adapts a bit based on the block. For example, perhaps the Group block's layout panel has the "inherit default layout" and "alignments" parts, as well as blockGap — while the Column's block layout panel has the columns count UI, stack on mobile, and blockGap. I'll relay these thoughts on my original issue, but wanted to chime in here.

In other new though — I tested the PR and it works great! :)

apeatling reviewed (commented):
Tested this and it's working as expected for me. I also agree on thinking about where these settings go.

I'm also concerned about calling this "Block gap" as a setting in this context. I know with the supports mechanism this is difficult, but would it not be better named contextually in situations like this? So calling it "Column gap" here? Not a blocker obviously.

andrewserong commented:
Thanks for testing @apeatling!

> but would it not be better named contextually in situations like this?

That's a great point — I suspect if I were a new user, I'd have no idea what "Block gap" means, but "Column gap" or ("Image gap" in the context of a Gallery block) would be much clearer. I think it'd also help a bit with the "Block gap" control feeling possibly out of place in the Dimensions panel as it is now. With a better name, maybe it'll feel slightly more at home.

In some of the other designs folks have shared (e.g. Joen in this comment https://github.com/WordPress/gutenberg/issues/28356#issuecomment-915072804) I see that the label "Block spacing" has been used, which might be a bit more natural for users?

I really like the suggestion of "Column gap", though, so my preferred option would be to see if we can allow each block that opts-in to the blockGap support to set its own override label (@aaronrobertshaw had a good idea of seeing if we can set this in the `theme.json`). I think getting the labelling right will be important to the usability of the feature, so I'll dig in a little further and explore out options. Thanks!

ramonjd commented:
Nice work!

> To allow the user to tweak this, the margin control is added, so that the margin can be controlled separately to the gap on inner blocks.

I like the fact that margin is available if only to override the `margin-top: var( --wp--style--block-gap );`. I am wondering if we need this rule at all? The columns block stands on its own, as a single row of columns. With margin controls it seems superfluous. 

I'd expect a grid block, with multiple rows to spread spacing evenly.

Caveat: I'm probably missing context somewhere.

With margin support:

![with-margin-support](https://user-images.githubusercontent.com/6458278/133194094-3e1f205d-106c-4e90-a9a5-f5524e363ccb.gif)

Without:

![without-margin-support](https://user-images.githubusercontent.com/6458278/133194108-8e24af58-9328-4f7b-9e92-27958b43c54e.gif)

Do you think it's worth making margin a default control in the panel?

It's not a biggie. As others have mentioned it tests well otherwise!

>  I think getting the labelling right will be important to the usability of the feature, so I'll dig in a little further and explore out options. Thanks!

Providing the ability to tweak labels in theme.json sounds like a flexible solution. I ask myself though if it will open the doors to JSON bloat.

Gutter seems like a standard term to describe spacing between columns. I'm not sure if it's generic enough to speak to a wider audience though. 

It also might connote even distribution of space between column rows, which we won't have until we have a grid block 😆

andrewserong commented:
Thanks for testing @ramonjd!

> I like the fact that margin is available if only to override the `margin-top: var( --wp--style--block-gap );`. I am wondering if we need this rule at all?

I believe that margin rule comes from the Layout block support styling via #33812. There's some discussion from [this comment onwards](https://github.com/WordPress/gutenberg/pull/33812#issuecomment-891742157) on that PR that gives a bit more context as to the usage of top margins on child elements to control that particular spacing. As long as that CSS rule is present, I think it's worth us having the margin control in place (even if it's hidden in the menu) so that once we start adjusting the gap value, we can fix the margin in place if we need to. At least, that was my thinking!

> Providing the ability to tweak labels in theme.json sounds like a flexible solution. I ask myself though if it will open the doors to JSON bloat.

Quite possibly! I think the idea is that if we can have it in Gutenberg's `theme.json` for each of the blocks, then individual themes won't need to override it, so hopefully that bloat won't be something that theme authors need to contend with. But maybe `block.json` is a better place for it? Either way, I'm happy to explore our options. I might do that in a separate PR, though, since the solution will be applicable to each of the opt-ins rather than just the Columns block.

jasmussen commented:
> In some of the other designs folks have shared (e.g. Joen in this comment #28356 (comment)) I see that the label "Block spacing" has been used, which might be a bit more natural for users?

Block spacing is probably a great thing to start with, as it's agnostic across the blocks on which it's applied. At the moment, the columns block is decidedly desktop-first in its approach to a row of columns, but even now it has a "stack on mobile" option, where columns wrap onto new lines. In that context, and in light of potential intrinsic responsiveness (#34641), block spacing works to indicate it's meant both for vertical and horizontal spacing. "Column spacing" might work as well, but before we add extra conditionals it feels worth it to see if just "Block spacing" feels intuitive enough in use.

andrewserong committed:
Try setting a blockGap label using a context string, update fallback label to Block spacing

andrewserong commented:
Thanks for the feedback @jasmussen!

I was curious to see how we might go about overriding the label, because I was thinking about those other use cases for gap support that might not involve inner blocks (and to see how difficult it might be). I've updated this PR with an exploration:

It seemed to be fairly straightforward to add in a couple of simple functions to use the `__experimentalLabel` function that can be added to blocks when they're registered. By passing in a `blockGap` context string, we can return an override for the block. There were a couple of other ideas floated (e.g. use the `theme-i18n.json` file, or add a filter, [which was mentioned here](https://github.com/WordPress/gutenberg/pull/29155#issuecomment-806583346) by @oandregal ). Hooking into the label seemed to be an approach that required the least amount of code from within the block, but at the same time, @mtias [did mention on a previous PR](https://github.com/WordPress/gutenberg/pull/29155#issuecomment-802845248) that we should try to keep the supports as simple as possible, so I'm happy to revert if this approach doesn't seem viable or warranted!

To show what it looks like now, here's the view with the `Column spacing` override text, and with the `Block spacing` fallback for blocks that don't provide their own `__experimentalLabel` with the `blockGap` context string:

| Column spacing | Block spacing fallback |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/133386553-7d85a635-b067-4c97-8763-28266a379642.png) | ![image](https://user-images.githubusercontent.com/14988353/133386602-a74e8a1e-c72d-4903-b4b9-5aea44949220.png) |

One problem with `Column spacing` versus `Column gap` is that the former is _slightly_ too long for the single column in the global styles panel (since that panel has more padding than the block inspector sidebar). So if we go with that string, we might need to move the control to be a full width column to avoid truncating the string:

<img src="https://user-images.githubusercontent.com/14988353/133386943-adae7c57-5e53-4f8c-8ff0-9d47b26b23b6.png" width="300" />

I'm keen to hear what everyone thinks, both for which string to use, and for what's an appropriate approach for overriding a block support's control label from a block's settings. I'm not at all attached to the approach that I included in this PR, so happy to go in another direction or revert as needed 🙂

jasmussen commented:
Thanks for continuing on this. One instinct I have in exploring new ground is to always do as little as possible and leave as much as possible to followups. That way they can happen if they turn out to be necessary, whereas if we start with too much we don't realize what might have been unnecessary in the first place. In the case of gap, the precedence I keep coming back to is Figma's autolayout, which calls the feature "Spacing between items" — somewhat similarly agnostic to "Block spacing". So it's nice to know that tweaking the label is possible/easy, in case it becomes clear we really do need it. 

In the case of the available space, while the following mockup remains subject to change per details being worked on in #34574, it does point to the block spacing control always being a full row, rather than a 50% width column item:

<img width="1012" alt="Spacing" src="https://user-images.githubusercontent.com/1204802/133390309-63cc65b6-0315-47b4-9312-e1a5a757c993.png">

In the above, the icon to the right of the slider both serves to indicate which spacing is being adjusted, and as a button to toggle axial controls. Followup territory of course, but it does suggest the full row of space is available.

youknowriad commented on packages/block-library/src/columns/index.js:
```diff
@@ -79,6 +79,11 @@ export const settings = {
 			},
 		],
 	},
+	__experimentalLabel( _, { context } ) {
```
This API sounds very weird to me, using a "block API" to label a block support. Did we consider a config in the block.json file?

andrewserong replied:
It is pretty weird, particularly now that I've had a chance to sleep on it! I think a config in `block.json` would be a better way to go about it.

I agree with Joen, though, that it'd be good to try to do the least amount of work here as possible for the Columns block, so in the immediate term, I'm going to revert the label overriding and just set the string to a hard-coded `Block spacing` for the time being.

If and when we get back to it, I'll take a look at seeing if we can set it in the `block.json` file. Thanks for taking a look 🙂

andrewserong committed:
Revert label override, but retain update to default string of Block spacing

andrewserong commented:
Thanks for the feedback @jasmussen!

> One instinct I have in exploring new ground is to always do as little as possible and leave as much as possible to followups. That way they can happen if they turn out to be necessary, whereas if we start with too much we don't realize what might have been unnecessary in the first place.

That's very wise, and I agree, I really like small iterative PRs rather than packing in too much at once. I've reverted the label override behaviour, which we can defer to figuring out further down the track if we need it. I've retained the update to the string so that it now reads `Block spacing`.

> it does point to the block spacing control always being a full row, rather than a 50% width column item

Good to know! Since the label `Block spacing` fits within the single 50% width column, I've left that styling as-is so that the UnitControl is a comfortable width, rather than stretching on too long. I thought we could take a look at switching to full width when we update the control that we're using when we eventually implement axial controls?

I explored removing the `single-column` class, but we'd then need to add some CSS to force the UnitControl to an appropriate width, so I left it in place, but here's a screenshot for visibility sake 🙂

| Removing `single-column` class | Current state of this PR |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/133538805-f364ec32-7afe-4f7a-83b2-fde50a53b5a8.png) | ![image](https://user-images.githubusercontent.com/14988353/133538772-1af09f69-0374-44ca-803f-2221c1491150.png) |

Do let me know if you'd like further changes or tweaks, of course!

jasmussen commented:
Knowing this is an area and control we'll come back to, seems fine. Worth testing in german as that usually adds 20% width.

andrewserong commented:
> Worth testing in german as that usually adds 20% width.

Good thinking. If Google Translate is anything to go by, the German translation is `Blockabstand` which fits in nicely (just via inspect element since we haven't gotten this translated yet):

<img src="https://user-images.githubusercontent.com/14988353/133729537-1cdd10bc-fe6d-4c12-884c-1926afc04261.png" width="300" />

Though CC @ramonjd on the German translation to check that it's actually the right word 😄

andrewserong committed:
Remove single-column class, add 80px width to UnitControls

andrewserong commented:
I've just given this a rebase to keep it current and test with the new navigation in global styles. I've also removed the `single-column` class on the block spacing/gap row in the dimensions panel, and added in `80px` width to the UnitControl, which seems to be fairly consistent with some of the other UnitControl components we're using in various places in the UI (happy to change it of course!). Here's a couple of screenshots of the current state of the PR:

| Global styles | Post editor sidebar |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/135391049-9ac7c15a-2912-4ed7-aa95-0655c66898f5.png) | ![image](https://user-images.githubusercontent.com/14988353/135391059-0bdf3ce8-d0eb-476a-949a-5f4fc2c3556f.png) |

@jasmussen I'm aware the cut-off date for getting features into 5.9 is approaching. Do you think it'd be worth us getting this feature in with it sitting in the Dimensions panel (knowing that we'll likely then move it to another place in the following version), or would you prefer to continue the discussion / design iterations of the layout / spacing panels before we progress this further?

ramonjd commented:
> Though CC @ramonjd on the German translation to check that it's actually the right word 😄

Sorry, I just saw this @andrewserong "Abstand" is indeed the right noun I'm pretty sure, so "Blockabstand" is 👍

jasmussen commented:
>  I'm aware the cut-off date for getting features into 5.9 is approaching. Do you think it'd be worth us getting this feature in with it sitting in the Dimensions panel (knowing that we'll likely then move it to another place in the following version), or would you prefer to continue the discussion / design iterations of the layout / spacing panels before we progress this further?

I'm personally a fan of both `gap`, and okay with moving things around a little bit in the future. But just because of the wide usage of the columns block, including in patterns, it might be good to both keep testing it widely, and get a sanity check from @apeatling and/or @youknowriad.

youknowriad reviewed (commented):
I think personally we should consider merging this to be able to test it properly in the Gutenberg plugin before 5.9.

apeatling commented:
I agree with Riad, let's test in the plugin first for a cycle.

ramonjd reviewed (approved):
The columns spacing + margin looks good to me. Copy change is also 👍 



https://user-images.githubusercontent.com/6458278/135946042-4612a1c9-9c9f-4f88-b3a5-267d38da017e.mp4






✅ I can apply block spacing and margins in the block and site editors. The frontend appears as I'd expect and the styles are applied, e.g, 

```
.wp-block-columns {
    margin-top: 62px;
    margin-bottom: 62px;
    --wp--style--block-gap: 64px;
}
```

I noticed that I cannot completely clear the margin value from the field. 

I've approved this PR as the changes aren't related it appears.

In the Block Editor the value persists until I remove the field or Reset All.

In the Site Editor it persists even after I've Reset All.


![columns-margin-reset](https://user-images.githubusercontent.com/6458278/135945895-c60dfbbc-75c7-42e5-ad7e-d169d77f678b.gif)

Is that expected?

ramonjd commented:
> I noticed that I cannot completely clear the margin value from the field.

I was just testing this again and it's only when I clear the field and blur out. If I hit the Enter key, the value persists. Maybe we need to handle null values in the blur event of UnitControl? I don't know yet 🤷

andrewserong commented:
Thanks for testing @ramonjd! I think the different behaviour you're seeing with the `margin` support is that the BoxControl component that it uses sets [isPressEnterToChange](https://github.com/wordpress/gutenberg/blob/b5e0e421af1890577b3b45cff42d49610e1d82d9/packages/components/src/box-control/unit-control.js#L41-L41) on its UnitControl, so to clear the value, you need to hit Enter. For the gap control, where we're using the UnitControl component directly, we haven't set that flag.

I think I have a personal preference to not requiring pressing Enter to commit the value in the BoxControl, but that's probably something for us to look at in a separate PR? I can see an argument for going either way.

ramonjd commented:
> PressEnterToChange on its UnitControl, so to clear the value, you need to hit Enter. For the gap control, where we're using the UnitControl component directly, we haven't set that flag.

Thanks for confirming!  I was looking at [mayUpdateUnit()](https://github.com/WordPress/gutenberg/blob/14d80aa66819682118a73f3cb40485806b9b8c44/packages/components/src/unit-control/index.js#L86) and came across it. I wanted to see if we could do something differently `onBlur`, but yeah... another PR :)

> I think I have a personal preference to not requiring pressing Enter to commit the value in the BoxControl, but that's probably something for us to look at in a separate PR? I can see an argument for going either way.

Definitely. I just wanted to flag it.

I also have a personal preference (or maybe an expectation) that I can clear a form field's value this way, especially if the fields in the same panel behave differently.


Not a blocker for this PR. ✅

andrewserong commented:
> Not a blocker for this PR. ✅

Thanks for confirming 👍 

> I also have a personal preference (or maybe an expectation) that I can clear a form field's value this way, especially if the fields in the same panel behave differently.

Okay, great, since we both have the same expectation, I'm happy to look into a PR for the margin and padding controls so we can start the discussion of switching it off there.

I'll merge this PR in now. If anyone thinks of other follow-ups we should look at, too, please let me know!

brylie commented:
We sometimes need to space elements away from the edges of their parent container, such as when using a full-width Gutenberg block. Will the user also be able to add `left` and `right` margins?

E.g., the mockup in issue https://github.com/WordPress/gutenberg/issues/24874

![image](https://user-images.githubusercontent.com/17307/141437712-8f9f2729-b03e-4e1e-8b3c-c6e0b6e2199a.png)

![image](https://user-images.githubusercontent.com/17307/141438476-d1d0c242-cbbd-43ae-a581-12aaf84d73df.png)

ramonjd commented:
> Will the user also be able to add left and right margins?

Is this the effect that you're describing?

![Nov-15-2021 07-17-27](https://user-images.githubusercontent.com/6458278/141697151-1d2bf784-268b-4c73-a170-ee402e8aef33.gif)

To achieve the above effect, it's possible to turn left and right margin and padding support on in `theme.json`, e.g.,

```json
		"spacing": {
			"blockGap": true,
			"margin": true,
			"padding": false,
			"units": [ "px", "em", "rem", "vh", "vw", "%" ]
		},
```

I guess it depends on the  underlying layout styles to whether left and right margins will have any effect however.

For example, the default layout width carries with it the following CSS:

```css

    max-width: 650px;
    margin-left: auto !important;
    margin-right: auto !important;

```

<img width="253" alt="Screen Shot 2021-11-15 at 7 32 27 am" src="https://user-images.githubusercontent.com/6458278/141697574-50dc36f4-1f64-46ca-82d4-faa461c7cf76.png">

brylie commented:
> Is this the effect that you're describing?

Yes! Thanks for the examples too :smiley:

MuluhGodson commented:
@ramonjd 
I added the settings for custom paddings and margins: 
```json
"settings": {
		"spacing": {
			"customPadding": true,
			"customMargin": true,
			"units": [ "px", "em", "rem", "vh", "vw" ]
		}
	}
```

![image](https://user-images.githubusercontent.com/40151808/143198895-f140d164-7fb2-4480-a769-3307398dbfe6.png)

The padding controls display and works correctly. However, I can't see the margin controls. Is there some additional setup to be done?

P.S: I used `theme.json` v2 syntax but still the same

brylie commented:
@MuluhGodson, for clarification, what type of Gutenberg block settings do we see in the example above? It may be that padding is enabled for some blocks but not others?

MuluhGodson commented:
> @MuluhGodson, for clarification, what type of Gutenberg block settings do we see in the example above? It may be that padding is enabled for some blocks but not others?

The settings above are applied globally to all block elements. So each block can have its own padding and margin controls.

ramonjd commented:
> However, I can't see the margin controls. Is there some additional setup to be done?

👋 @MuluhGodson Which block are you trying to target, and where is the JSON?

I think @brylie has the answer: _"It may be that padding is enabled for some blocks but not others?"_

Enabling margin and padding globally via `theme.json` will "enable" margin and padding to used so long as block opt into them using the "supports" object. More info [here](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#spacing) on that.

So in your theme's `theme.json` you might have something like this under the "settings" object:

```json
		"spacing": {
			"blockGap": true,
			"margin": true,
			"padding": true,
			"units": [ "px", "em", "rem", "vh", "vw", "%" ]
		},
```

This "enables" all spacing supports, and the blocks that opt into them via their individual block.json settings will show them. If the block doesn't opt in

You can control individual block settings via `theme.json` as well by adding an entry to the "blocks" object.

So you could set the spacing controls to show for the core Group Block in `theme.json` like this (so long as you have the global settings enabled):

```json
		"blocks": {
			"core/group": {
				"spacing": {
					"blockGap": true,
					"margin": true,
					"padding": true,
					"units": [ "px", "em", "rem", "vh", "vw", "%" ]
				},
			}
		}
```

More information on theme.json over [here](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/).

Hope that helps!

brylie commented:
Should we be using `margin` / `padding` or `customMargin` / `customPadding` properties? The [block-supports spacing documentation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#spacing) uses `margin` and `padding`, while `customMargin` seems to also work (but not `customPadding` on whichever block type  is depicted above).

ramonjd commented:
> Should we be using margin / padding or customMargin / customPadding properties? 

Since https://github.com/WordPress/gutenberg/pull/36155 `margin` and `padding` are preferred, though it looks like there's backwards compatibility for the previous `custom*` properties.

MuluhGodson commented:
> Should we be using `margin` / `padding` or `customMargin` / `customPadding` properties? The [block-supports spacing documentation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#spacing) uses `margin` and `padding`, while `customMargin` seems to also work (but not `customPadding` on whichever block type is depicted above).

`theme.json` v1 using customPadding / customMargin while v2 removes the `custom` leaving is as `margin` / `padding` however there's backwards compatibility to v1 syntax. I will switch to v2.

mjamesderocher commented:
Is there a way to limit the margin to predefined options? So instead of allowing any size value, there would be choices of say, small, medium, and large that are set in theme.json? This is how text sizing works, and with text, you have the choice to enter a custom value, but can also disable custom values being entered in theme.json.

ramonjd commented:
> Is there a way to limit the margin to predefined options? So instead of allowing any size value, there would be choices of say, small, medium, and large that are set in theme.json? 

👋 This relates to the discussion over at https://github.com/WordPress/gutenberg/issues/38998 about standardized tokens.
